### ISSUE TITLE
Unify with bytestring Builder

### ISSUE THREAD

**Issue by aristidb** [meiersi/blaze-builder#17]
Are there plans to unify blaze-builder with bytestring builder?

**Comment by alevy** [comment]
+1

**Comment by chrisdone** [comment]
Is there a migration path to bytestring-builder?

**Comment by meiersi** [comment]
Yes, there is. @lpsmith has done most of the work to provide a compatibility release with his https://github.com/lpsmith/bytestring-builder library and the compatibility branch in this repo. It is up to me to finally release version 0.4 of blaze-builder that makes use of this good work.

**Comment by aristidb** [comment]
Sounds good!

Simon Meier notifications@github.com schrieb am Tue Jan 27 2015 at
19:43:00:

**Comment by chrisdone** [comment]
That sounds great, @meiersi. I look forward to it.

**Comment by lpsmith** [comment]
Ok,  I just sent a pull request to Simon for the latest version of my compat work.    it does pass the old test suite,  but the test suite is not comprehensive and it could be tested better.    IIRC,   [this buglet](https://github.com/lpsmith/blaze-builder/commit/066f81840279b526bf4f819fa64fd3e5a9ade7c6) passed the suite,   but was caught upon inspection of the output of some manual ad-hoc tests.

In any case,  the existing work should be pretty close to correct.   I am using this compatibility layer in a pretty minimal way in production,   where I have a daemon that's using the new builder that also touches a postgres database via `postgresql-simple`,   which has `blaze-builder` as part of it's public API.    However,  what it's doing via the compatibility layer is very simple so I wouldn't consider it that strong of evidence of correctness either.   In particular,  I'm pretty sure that buglet would also not have impacted that daemon.

I'm also of the opinion that,  as long as the programmers put a reasonable and honest effort into making it correct,  that releasing the software is an effective means of exposing any lingering problems.   

`snap` depends on the old internal modules,  which are no longer available in the compatibility layer.   I looked at getting snap running on the new builder,  but it always turned out to be a bit more effort than I was willing to put into it at that point in time.

**Comment by lpsmith** [comment]
So, things that should be done before release:
1.   merge the newer parts of the master branch.   Some of the newer changes in `master` are no longer relevant,  but some are,  e.g. the `.travis.yml` file.
2.   Investigate the buglet I pointed out,  and add a regression test if necessary.
3.   Code review... I don't think that an adequate code review needs to be extremely detailed,  but the build parameters and in particular the dependency version constraints do need some work.

**Comment by meiersi** [comment]
Hi Leon,

thank you very much for this great work. I'll get to work on that and your pull request this Friday.

**Comment by meiersi** [comment]
I've looked at the code and I think there is one further issue that we should fix.
1. Deprecate the library for general use and point people to the bytestring builder.

**Comment by lpsmith** [comment]
Ok,  after a chat with Simon,  we decided that I'm going to take over maintainership of blaze-builder.    Also,  I finally released blaze-builder-0.4.0.0,  which is mostly just a wrapper around the new builder.

**Comment by meiersi** [comment]
Thanks a lot @lpsmith for taking over and handling this release!

2015-02-12 11:15 GMT+01:00 Leon P Smith notifications@github.com:

**Comment by chrisdone** [comment]
This sounds awesome.

So if I migrate my library to bytestring-builder, is there some `BS.Builder -> Blaze.Builder` function that users of my library can use? E.g. http-client [expects a `blaze-builder`](http://hackage.haskell.org/package/http-client-0.4.7.1/docs/Network-HTTP-Client.html#v:RequestBodyBuilder), and e.g. lucid would expose [a bytestring-builder](http://hackage.haskell.org/package/lucid-2.9.0/docs/Lucid.html#v:execHtmlT), so there'd need to be a (hopefully essentially no-op) conversion between the two types.

(Otherwise I can't reasonably upgrade to bytestring-builder without making my users unable to use my library with other libraries that expect blaze.)

**Comment by lpsmith** [comment]
@chrisdone `Blaze.Builder` is just a reexport of the `BS.Builder` type,  so the function you are looking for is `id`. ;-)

**Comment by chrisdone** [comment]
@lpsmith Okay, so I should wait until packages using blaze-builder have upgraded to this version before I can start using bytestring-builder.

**Comment by lpsmith** [comment]
@chrisdone,  if the packages you want to use only use the public interface,  the only thing you would need to do is (possibly) bump the upper bounds.   If they touch the internals,  like `snap` does for instance,  then yes,  you'll have to wait.

**Comment by lpsmith** [comment]
Although,  perhaps I should tweak the documentation a bit to clarify the issue that `Blaze.Builder` and `BS.Builder` are exactly the same type.

**Comment by chrisdone** [comment]
No problem, it seems that it'll be a straight-forward bump for package maintainers. This [stackage issue](https://github.com/fpco/stackage/issues/442) gives a good indication of what needs to be updated.

**Comment by lpsmith** [comment]
Well,  snap is a bit more involved,  but it still shouldn't be a serious issue.   I wrote up a [more detailed blog post](http://blog.melding-monads.com/2015/02/12/announcing-blaze-builder-0-4/) regarding this new release.
